ballista/rust/scheduler/src/state/persistent_state.rs

```diff
@@ -0,0 +1,312 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+use log::debug;
+use parking_lot::RwLock;
+use prost::Message;
+use std::any::type_name;
+use std::collections::HashMap;
+use std::sync::Arc;
+
+use ballista_core::error::{BallistaError, Result};
+
+use ballista_core::serde::protobuf::JobStatus;
+
+use crate::state::backend::StateBackendClient;
+use ballista_core::serde::scheduler::ExecutorMetadata;
+use ballista_core::serde::{protobuf, AsExecutionPlan, AsLogicalPlan, BallistaCodec};
+use datafusion::physical_plan::ExecutionPlan;
+use datafusion::prelude::ExecutionContext;
+
+type StageKey = (String, u32);
+
+#[derive(Clone)]
+pub(crate) struct PersistentSchedulerState<
+    T: 'static + AsLogicalPlan,
+    U: 'static + AsExecutionPlan,
+> {
+    // for db
+    config_client: Arc<dyn StateBackendClient>,
+    namespace: String,
+    pub(crate) codec: BallistaCodec<T, U>,
+
+    // for in-memory cache
+    executors_metadata: Arc<RwLock<HashMap<String, ExecutorMetadata>>>,
```

Comment:
Should executor metadata be persistent? Seems like it could be useful to speed up a scheduler restart since it would have to wait for executors to re-register but it seems like in the case of push-based scheduling we need to wait for heartbeats anyway and in the poll-based scheduling we wouldn't schedule work until the executor registers.

Reply:
Thanks @thinkharderdev for reviewing this PR. Actually the executor metadata is persistent to some backend storage. However, for fast reading, they are cached in memory. For all of those states in *PersistentSchedulerState*, firstly they will be persistent. Then they will be cached in memory.

Reply:
Right, that part makes sense. I just am wondering whether it makes sense to, on a scheduler restart, to reload the executor state from persistent storage rather than let the executors re-register. In the latter case, executor metadata could be volatile state.

Reply:
Suppose there's a cluster deployed on k8s in a standalone mode and we want to redeploy the scheduler. It would be better for the scheduler to reload the topology info from persistent storage. Whether the executor metadata be overdue or not, it should be managed by the executor heartbeat.